File: kattis/guessthedatastructure.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	int comm;
	while (cin >> comm) {
		// cout << "ini comm " << comm << endl;
		stack<int> s;
		queue<int> q;
		priority_queue<int> pq;
		vector<int> ori;
		vector<int> ss;
		vector<int> qs;
		vector<int> pqs;
		bool can = true;
		for (int i = 0; i < comm; i++) {
			int op, num;
			cin >> op >> num;
			if (op == 1) {
				s.push(num);
				q.push(num);
				pq.push(num);
			}
			else if (op == 2) {
				if (s.size() == 0) {
					can = false;
					ori.push_back(num);
				}
				else {
					ori.push_back(num);
					ss.push_back(s.top());
					s.pop();
					qs.push_back(q.front());
					q.pop();
					pqs.push_back(pq.top());
					pq.pop();			
				}
			}
		}
		int sim = 0;
		if (ori == ss) {
			sim++;
		}
		if (ori == qs) {
			sim++;
		}
		if (ori == pqs) {
			sim++;
		}
		if (sim == 0 || !can) {
			cout << "impossible" << endl;
		}
		else if (sim > 1) {
			cout << "not sure" << endl;
		}
		else if (sim == 1) {
			if (ori == ss) {
				cout << "stack" << endl;
			}
			else if (ori == qs) {
				cout << "queue" << endl;
			}
			else if (ori == pqs) {
				cout << "priority queue" << endl;
			}
		}
	}
}

File: kattis/guessthedatastructure.java
import java.util.*;

class Main {
	public static void main(String[] args) {
		Kattio io = new Kattio(System.in);

		while (io.hasMoreTokens()) {
			Stack<Integer> stk = new Stack<>();
			Queue<Integer> q = new LinkedList<>();
			PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
			int n = io.getInt();

			boolean is_s = true;
			boolean is_q = true;
			boolean is_pq = true;

			for (int i = 0; i < n; i++) {
				int command = io.getInt();
				if (command == 1) {
					int x = io.getInt();
					stk.push(x);
					q.add(x);
					pq.add(x);
				} else {
					int x = io.getInt();
					if (stk.isEmpty()) {
						is_s = false;
					} else {
						if (stk.peek() != x) {
							is_s = false;
						}
						
						stk.pop();
					} 

					if (q.isEmpty()) {
						is_q = false;
					} else {
						if (q.peek() != x) {
							is_q = false;
						}
						q.remove();
					}

					if (pq.isEmpty()) {
						is_pq = false;
					} else {
						if (pq.peek() != x) {
							is_pq = false;
						}
						pq.remove();
					}
				}
			}

			int score = (is_s ? 1 : 0) + (is_q ? 1 : 0) + (is_pq ? 1 : 0);
			if (score == 0) {
				io.println("impossible");
			} else if (score == 1) {
				if (is_s) {
					io.println("stack");
				} else if (is_q) {
					io.println("queue");
				} else {
					io.println("priority queue");
				}
			} else {
				io.println("not sure");
			}
		}

		io.flush();
	}
}